src/lib/logger/logger.ml

```diff
@@ -81,11 +81,160 @@ module Message = struct
+module Processor = struct
+  module type S = sig
+    type t
+
+    val process : t -> Message.t -> string option
+  end
+
+  type t = T : (module S with type t = 't) * 't -> t
+
+  module Raw = struct
+    type t = unit
+
+    let create () = ()
+
+    let process () msg = Some (Yojson.Safe.to_string (Message.to_yojson msg))
+  end
+
+  module Pretty = struct
+    type t = {log_level: Level.t; config: Logproc_lib.Interpolator.config}
+
+    let create ~log_level ~config = {log_level; config}
+
+    let process {log_level; config} msg =
+      let open Message in
+      if msg.level < log_level then None
+      else
+        match
+          Logproc_lib.Interpolator.interpolate config msg.message msg.metadata
+        with
+        | Error err ->
+            Core.printf "logproc interpolation error: %s\n" err ;
+            None
+        | Ok (str, _) ->
+            Some (Level.show msg.level ^ ": " ^ str)
+  end
+
+  let raw () = T ((module Raw), Raw.create ())
+
+  let pretty ~log_level ~config =
+    T ((module Pretty), Pretty.create ~log_level ~config)
+end
+
+module Transport = struct
+  module type S = sig
+    type t
+
+    val transport : t -> string -> unit
+  end
+
+  type t = T : (module S with type t = 't) * 't -> t
+
+  module Stdout = struct
+    type t = unit
+
+    let create () = ()
+
+    let transport () = Core.print_endline
+  end
+
+  let stdout () = T ((module Stdout), Stdout.create ())
+
+  module File_system = struct
+    module Dumb_logrotate = struct
+      open Core.Unix
+
+      (* 512MB *)
+      let max_size = 1024 * 1024 * 512
+
+      let log_perm = 0o644
+
+      let primary_log_name = "coda.log.0"
+
+      let secondary_log_name = "coda.log.1"
+
+      type t =
+        { directory: string
+        ; mutable primary_log: File_descr.t
+        ; mutable primary_log_size: int }
+
+      let create ~directory =
+        let primary_log_loc = Filename.concat directory primary_log_name in
+        let primary_log_size =
+          if Result.is_ok (access primary_log_loc [`Read; `Write]) then
+            let log_stats = stat primary_log_loc in
+            Int64.to_int_exn log_stats.st_size
+          else 0
+        in
+        let primary_log =
+          openfile ~perm:log_perm ~mode:[O_RDWR; O_APPEND] primary_log_loc
+        in
+        {directory; primary_log; primary_log_size}
+
+      let rotate t =
+        let primary_log_loc = Filename.concat t.directory primary_log_name in
+        let secondary_log_loc =
+          Filename.concat t.directory secondary_log_name
+        in
+        close t.primary_log ;
+        rename ~src:primary_log_loc ~dst:secondary_log_loc ;
+        t.primary_log
+        <- openfile ~perm:log_perm ~mode:[O_RDWR; O_TRUNC] primary_log_loc ;
+        t.primary_log_size <- 0
+
+      let transport t str =
+        if t.primary_log_size > max_size then rotate t ;
+        let len = String.length str in
+        if write t.primary_log ~buf:(Bytes.of_string str) ~len <> len then
```

Comment:
Using `async` would handle partial writes here... If you keep it to `Core.Unix` can you add a new issue for a `TODO` about handling the partial write?

Reply:
I will add an issue. I avoided using Async for now due to the increased complexity to the implementation (wouldn't be that bad if you think we should do it now rather than later, but it either involves some internally muxing of mutable deferreds or making the logger interface async, which I think is a bad idea).